DQM/HLTEvF/plugins/LumiMonitor.cc

```diff
@@ -55,43 +54,42 @@ class LumiMonitor : public DQMGlobalEDAnalyzer<Histograms> {
   static MEbinning getHistoPSet(const edm::ParameterSet& pset);
   static MEbinning getHistoLSPSet(const edm::ParameterSet& pset);
 
-  edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> const trkTopoToken_;
-
-  std::string folderName_;
-
-  edm::EDGetTokenT<LumiScalersCollection> lumiScalersToken_;
-  MEbinning lumi_binning_;
-  MEbinning pu_binning_;
-  MEbinning ls_binning_;
-
-  bool doPixelLumi_;
-  edm::EDGetTokenT<edmNew::DetSetVector<SiPixelCluster>> pixelClustersToken_;
-  bool useBPixLayer1_;
-  int minNumberOfPixelsPerCluster_;
-  float minPixelClusterCharge_;
-  MEbinning pixelCluster_binning_;
-  MEbinning pixellumi_binning_;
+  std::string const folderName_;
 
-  edm::EDGetTokenT<LumiSummary> lumiSummaryToken_;
+  edm::EDGetTokenT<LumiScalersCollection> const lumiScalersToken_;
+  edm::EDGetTokenT<OnlineLuminosityRecord> const onlineMetaDataDigisToken_;
+  MEbinning const lumi_binning_;
+  MEbinning const pu_binning_;
+  MEbinning const ls_binning_;
 
-  float lumi_factor_per_bx_;
+  bool const doPixelLumi_;
+  edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> const trkTopoToken_;
+  edm::EDGetTokenT<edmNew::DetSetVector<SiPixelCluster>> const pixelClustersToken_;
+  bool const useBPixLayer1_;
+  int const minNumberOfPixelsPerCluster_;
+  float const minPixelClusterCharge_;
+  MEbinning const pixelCluster_binning_;
+  MEbinning const pixellumi_binning_;
+  float const lumi_factor_per_bx_;
 };
 
 // -----------------------------
 //  constructors and destructor
 // -----------------------------
 
 LumiMonitor::LumiMonitor(const edm::ParameterSet& config)
-    : trkTopoToken_{esConsumes()},
-      folderName_(config.getParameter<std::string>("FolderName")),
-      lumiScalersToken_(consumes<LumiScalersCollection>(config.getParameter<edm::InputTag>("scalers"))),
+    : folderName_(config.getParameter<std::string>("FolderName")),
+      lumiScalersToken_(consumes(config.getParameter<edm::InputTag>("scalers"))),
+      onlineMetaDataDigisToken_(consumes(config.getParameter<edm::InputTag>("onlineMetaDataDigis"))),
       lumi_binning_(getHistoPSet(
           config.getParameter<edm::ParameterSet>("histoPSet").getParameter<edm::ParameterSet>("lumiPSet"))),
       pu_binning_(
           getHistoPSet(config.getParameter<edm::ParameterSet>("histoPSet").getParameter<edm::ParameterSet>("puPSet"))),
       ls_binning_(getHistoLSPSet(
           config.getParameter<edm::ParameterSet>("histoPSet").getParameter<edm::ParameterSet>("lsPSet"))),
       doPixelLumi_(config.getParameter<bool>("doPixelLumi")),
+      trkTopoToken_(doPixelLumi_ ? esConsumes<TrackerTopology, TrackerTopologyRcd>()
+                                 : edm::ESGetToken<TrackerTopology, TrackerTopologyRcd>()),
```

Comment:
This is meant to avoid consuming the record when the plugin does not make use of it.

The following would be slightly more compact
```suggestion
      trkTopoToken_(doPixelLumi_ ? esConsumes<TrackerTopology, TrackerTopologyRcd>() : trkTopoToken_),
```
but I haven't seen this approach elsewhere, and I don't know if this is a good idea.

The even-more-compact version
```suggestion
      trkTopoToken_(doPixelLumi_ ? esConsumes() : trkTopoToken_),
```
unfortunately didn't compile:
```
error: operands to '?:' have different types 'edm::EDConsumerBaseESAdaptor<edm::Transition::Event>' and 'edm::ESGetToken<TrackerTopology, TrackerTopologyRcd>'
   91 |       trkTopoToken_(doPixelLumi_ ? esConsumes()
      |                     ~~~~~~~~~~~~~^~~~~~~~~~~~~~
   92 |                                  : edm::ESGetToken<TrackerTopology, TrackerTopologyRcd>()),
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
As an aside: if there is feedback on this technical point, I'd be interested to know (doing `x_{bool ? a : x_}` in the initialiser list, and the fact that one of the options doesn't compile, which is probably expected, but I'm wondering if there is a way to make it work).

Reply:
This does not seem like a good idea, because `x_` would end up uninitialised ?

Simple example using integers and strings:
```c++
#include <iostream>
#include <string>

struct Data {
public:
  Data(int x, std::string s) : x_{x}, s_{std::move(s)}
  {}

  Data() : x_{x_}, s_{s_}
  {}

  int x() const {
    return x_;
  }

  std::string const& s() const {
    return s_;
  }

private:
  int x_;
  std::string s_;
};

int main(void) {

  Data value{42, "answer"};
  std::cout << value.s() << " is " << value.x() << std::endl;

  Data data;
  std::cout << data.s() << " is " << data.x() << std::endl;

  return 0;
}
```

The compiler warns that `x_` is initialised:
```
$ g++ -std=c++17 -O3 -flto -g -Wall -fPIC -MMD -march=native -mtune=native  -c test.cc -o test.cc.o
test.cc: In member function ‘Data::Data()’:
test.cc:9:15: warning: ‘this_6(D)->x_’ is used uninitialized [-Wuninitialized]
    9 |   Data() : x_{x_}, s_{s_}
      |               ^~
```

Worse, there is no warning about `s_` being uninitialised, but I get a crash:
```
$ ./test 
answer is 42
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted (core dumped)
```

Reply:
Indeed.. thanks the example, it's clearer now!